test/IRGen/extern_c.swift

```diff
@@ -0,0 +1,37 @@
+// RUN: %empty-directory(%t)
+// RUN: split-file %s %t
+// RUN: %target-swift-frontend -emit-ir %t/extern_c.swift -I%t | %FileCheck %s
+
+//--- c_types.h
+struct c_struct {
+  int field;
+};
+
+//--- module.modulemap
+module c_types {
+  header "c_types.h"
+}
+
+//--- extern_c.swift
+import c_types
+
+func test() {
+  // CHECK: call void @explicit_extern_c()
+  explicit_extern_c()
+  // CHECK: call void @implicit_extern_c()
+  implicit_extern_c()
+
+  // CHECK: call i32 @"+"({{.*}})
+  _ = c_struct(field: 1) + c_struct(field: 2)
+}
+
+test()
+
+// CHECK: declare void @explicit_extern_c()
+@_extern(c, "explicit_extern_c") func explicit_extern_c()
+
+// CHECK: declare void @implicit_extern_c()
+@_extern(c) func implicit_extern_c()
+
+// CHECK: declare i32 @"+"({{.*}})
+@_extern(c) func +(a: c_struct, b: c_struct) -> c_struct
```

Comment:
I guess we don't check that the name is a valid C symbol name? Maybe we should? You wouldn't be allowed to declare a function called `+` in C.

Reply:
Yeah, good catching. "Valid" C symbol is actually implementation-defined thing in the C standard spec considering Universal Character Names, as far as I understand. I took a look in Clang to find a way to check if a string is a valid C identifier in Clang, but I couldn't find it. 
(there is `isValidAsciiIdentifier` but it does not support UCNs, and it seems any identifiers that can be lex'ed are treated as valid)

So there are two options here. 1. Accept only ASCII identifiers using [`clang::isValidAsciiIdentifier`](https://github.com/llvm/llvm-project/blob/761c9dd92789b744607dc9c8c5071fef340fd86f/clang/include/clang/Basic/CharInfo.h#L224-L238), 2. Accept any identifiers that are valid as Swift function name. 

Option 1 can reject identifiers including UCNs but can reject apparently invalid C identifiers. Option 2 accepts all valid C identifiers, but can accept invalid identifiers too.

I think Option 1 is conservative, but better approach to be practical and we can incrementally unlock UCNs things later. WDYT?

Reply:
Agreed, I would go with option 1 for now. UCNs and unicode in identifiers more generally are not particularly widely used in C, and I think they're especially unlikely in situations where we'd like to use `@_extern(c)` (because that's mainly going to be used to call functions in the runtime and/or legacy C code, much of which predates Unicode identifier support anyway).

Reply:
What we _could_ do is make it so that using a potentially invalid name is just a warning, rather than an error per se. That way people get told that something might go wrong, and if they feel they know better then they can carry on anyway? I'm not sure how I feel about that, but it's worth thinking about.

Reply:
Considering this feature is for-expert, I think it makes sense to make it just a warning not to block anything that *might* be valid.

Reply:
Maybe we only want the warning if the name is implicit. For explicit names, as you say, this is an "expert" feature, so if I _really_, _really_ want to call a function whose symbol is "+" or even "🥸", maybe it should just trust me at that point.

Reply:
Ah, it seems the right way! Thanks!